### PR TITLE
Extend /v1/parties endpoint to support fetch by party IDs

### LINKED ISSUE #4512: include parties' display names for signatories, observers when present
Where parties occur in the domain model (that is, _not_ in LF values such as contract payloads), we can supply their display names alongside, like

```ts
{identifier: string,
 displayName?: string}
```

One presumes that the observers and signatories of a contract ought to have access to the display names of the others.

Example event:

```json
{
    "created": {
        "observers": [],
        "agreementText": "",
        "payload": {
            "observers": [],
            "issuer": "Alice",
            "amount": "999.99",
            "currency": "USD",
            "owner": "Alice"
        },
        "signatories": [
            {
                "identifier": "Alice",
                "displayName": "Alice Baker"
            }
        ],
        "contractId": "#1:0",
        "templateId": "b70bbfbc77a4790f66d4840cb19f657dd20848f5e2f64e39ad404a6cbd98cf75:Iou:Iou"
    }
}
```

### PR DESCRIPTION
**Pull request by leo-da**
Closes #4512

- `GET` returns all known parties
- `POST` returns only parties specified in the input array, unknown parties reported as `warnings`, in the format consistent with `unknownTemplateIds`
- `POST` returns BadRequest(400) error if input array of party identifiers is empty: `[]`.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: docs/source/json-api/index.rst [hunk 1]
```diff
@@ -803,12 +803,19 @@ Nonempty HTTP Response with Unknown Template IDs Warning
...
-- Content: <EMPTY>
+- Method: ``POST``
+- Content-Type: ``application/json``
+- Content:
+
+.. code-block:: json
+
+    ["Alice", "Bob", "Dave"]
+
+If empty JSON array is passed: ``[]``, this endpoint will return all known parties.
```
Comment (hurryabit) [inline docs/source/json-api/index.rst]: I'm not convinced that is a good idea. I prefer staying consistent and returning an empty array in that situation. If we want to be able to get all parties on the POST endpoint, we should signal this by sending no content at all. But do we actually want this?

File: docs/source/json-api/index.rst [hunk 2]
```diff
@@ -819,15 +826,47 @@ HTTP Response
...
+          "isLocal": true
+        },
+        {
+          "identifier": "Dave",
+          "isLocal": true
+        }
+      ]
     }
 
+Please note that the order of the party objects in the response is not guaranteed to match the order of the passed party identifiers.
```
Comment (hurryabit) [inline docs/source/json-api/index.rst]: What happens if there are some known parties and some unknown parties? I expect that we only receive records for the known parties. We should consider adding a `warnings` field containing the unknown parties, but I'm fine doing this in a separate PR.

File: docs/source/json-api/index.rst [hunk 3]
```diff
@@ -819,15 +826,47 @@ HTTP Response
...
+
+An empty JSON array will be returned in the ``result`` element of the response if passed identifiers did not match any party.
```
Comment (hurryabit) [inline docs/source/json-api/index.rst]: If we clarify what happens with unknown parties above, this should not be needed anymore.

File: docs/source/json-api/index.rst (hunk 1 above)
Comment (leo-da) [reply docs/source/json-api/index.rst]: all right, we will return only parties specified in the input. Empty input array would mean empty array in the output.

File: docs/source/json-api/index.rst (hunk 2 above)
Comment (leo-da) [reply docs/source/json-api/index.rst]: I will add warnings in the follow up PR

Comment (leo-da) [reply docs/source/json-api/index.rst]: scratch the above it is a rather small change, easier to add it to this PR and update the docs once.

Comment (leo-da) [reply docs/source/json-api/index.rst]: scratch the above it is a rather small change, easier to add it to this PR and update the docs once.

File: docs/source/json-api/index.rst (hunk 1 above)
Comment (leo-da) [reply docs/source/json-api/index.rst]: we actually expect a NonEmptyList of party ids now, if empty array passed, the current implementation returns a documented error:
```
{
  "status": 400,
  "errors": [
    "JsonReaderError. Cannot read JSON: <[]>. Cause: spray.json.DeserializationException: must be a list with at least 1 element"
  ]
}
```
I think error makes more sense, why would you ask for an empty array.

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/util/Collections.scala [hunk 4]
```diff
@@ -31,4 +31,7 @@ object Collections {
     def collect[B](f: A PartialFunction B): Option[NonEmptyList[B]] =
       self.list.collect(f).toNel
   }
+
+  def toNonEmptySet[A](as: NonEmptyList[A]): OneAnd[Set, A] =
+    OneAnd(as.head, as.tail.foldLeft(Set.empty[A])(_ + _) - as.head)
```
Comment (S11001001) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/util/Collections.scala]:
```suggestion
    import scalaz.syntax.foldable._
    OneAnd(as.head, as.tail.toSet - as.head)
```

File: ledger-service/http-json/src/test/scala/com/digitalasset/http/AbstractHttpServiceIntegrationTest.scala [hunk 5]
```diff
@@ -83,11 +83,17 @@ trait AbstractHttpServiceIntegrationTestFuns extends StrictLogging {
...
-  protected def withHttpService[A]
-    : ((Uri, DomainJsonEncoder, DomainJsonDecoder) => Future[A]) => Future[A] =
+  protected def withHttpServiceAndClient[A]
+    : ((Uri, DomainJsonEncoder, DomainJsonDecoder, LedgerClient) => Future[A]) => Future[A] =
     HttpServiceTestFixture
       .withHttpService[A](testId, List(dar1, dar2), jdbcConfig, staticContentConfig)
 
+  protected def withHttpService[A](
+      f3: (Uri, DomainJsonEncoder, DomainJsonDecoder) => Future[A]): Future[A] =
+    HttpServiceTestFixture
```
Comment (S11001001) [inline ledger-service/http-json/src/test/scala/com/digitalasset/http/AbstractHttpServiceIntegrationTest.scala]:
```suggestion
    withHttpServiceAndClient((a, b, c, _) => f(a, b, c))
```

File: docs/source/json-api/index.rst (hunk 1 above)
Comment (hurryabit) [reply docs/source/json-api/index.rst]: Erroring out on the empty list is fine too.

**Pull request merged**